source/opt/debug_info_manager.h

```diff
@@ -0,0 +1,132 @@
+  // |new_var_id|. Return the new DebugDeclare or DebugValue. If this is the
+  // first DebugDeclare or DebugValue, keep it in |local_var_id_to_dbgdecl_|.
+  Instruction* CloneDebugDeclare(uint32_t orig_var_id, uint32_t new_var_id);
+
+  // Returns id of new DebugInlinedAt. Its Line operand is the line number
+  // of |line| if |line| is not nullptr. Otherwise, its Line operand
+  // is the line number of lexical scope of |scope|. Its Scope and Inlined
+  // operands are Scope and Inlined of |scope|. Note that this function puts
+  // the new DebugInlinedAt into the tail of debug instructions.
+  uint32_t CreateDebugInlinedAt(const Instruction* line,
+                                const DebugScope& scope);
+
+  // Creates a new DebugInfoNone instruction and returns it. In addition,
+  // insert the new DebugInfoNone instruction before the head of debug
+  // instructions.
+  Instruction* CreateDebugInfoNone();
+
+  // Returns DebugInlinedAt whose id is |dbg_inlined_at_id|. If it does not
+  // exist or it is not a DebugInlinedAt instruction, return nullptr.
+  Instruction* GetDebugInlinedAt(uint32_t dbg_inlined_at_id);
+
+  // Returns DebugFunction whose Function operand is |fn_id|. If it does not
+  // exist, return nullptr.
+  Instruction* GetDebugFunction(uint32_t fn_id) {
+    auto dbg_fn_it = fn_id_to_dbg_fn_.find(fn_id);
+    return dbg_fn_it == fn_id_to_dbg_fn_.end() ? nullptr : dbg_fn_it->second;
+  }
+
+  // Clones DebugInlinedAt whose id is |dbg_inlined_at_id|. If
+  // |dbg_inlined_at_id| is not an id of DebugInlinedAt, returns nullptr.
+  // Note that this function does not insert the new DebugInlinedAt into
+  // debug instruction list of the module.
+  Instruction* CloneDebugInlinedAt(uint32_t dbg_inlined_at_id);
+
+ private:
+  IRContext* context() { return context_; }
+
+  // Analyzes OpenCL.DebugInfo.100 instructions in the given |module| and
+  // populates data structures in this class.
+  void AnalyzeDebugInsts(Module& module);
+
+  // Returns the DebugDeclare instruction that corresponds to the variable
+  // with id |var_id|. Returns |nullptr| if one does not exists.
+  Instruction* GetDbgDeclareForVar(uint32_t var_id);
+
+  // Registers the DebugDeclare instruction |inst| into
+  // |local_var_id_to_dbgdecl_| using the variable operand of |inst| as a key.
+  // If |local_var_id_to_dbgdecl_| already has a key that corresponds to the
+  // variable operand of |inst|, it just returns without doing anything.
+  void RegisterDbgDeclareForVar(Instruction* inst);
+
+  // Returns the debug instruction whose id is |id|. Returns |nullptr| if one
+  // does not exists.
+  Instruction* GetDbgInst(uint32_t id);
+
+  // Registers the debug instruction |inst| into |id_to_dbg_inst_| using id of
+  // |inst| as a key.
+  void RegisterDbgInst(Instruction* inst);
+
+  // Registers the DebugFunction |inst| into |fn_id_to_dbg_fn_| using the
+  // function operand of |inst| as a key.
```

Comment:
Any special cases that need to be documented?  What happens if we have two debug function instructions for the same function?

Reply:
I am also worried about the case. I added an `assert()` in the function that crashes if the function was already registered with a different `DebugFunction`.

I am not sure what comment it must be but I explained it "If the function operand of |inst| is already registered with a different DebugFunction, it will result in an assertion failure"

Reply:
Or just say "The function referenced in |inst| must not already be registered."  don't mention the assert because assert are compiled out in release builds.